config/stored_requests.go

```diff
@@ -38,47 +39,84 @@ type PostgresConfig struct {
 	// It is a Template, rather than a full Query, because a single HTTP request may reference multiple Stored Requests.
 	//
 	// In the simplest case, this could be something like:
-	//   SELECT id, requestData FROM stored_requests WHERE id in %ID_LIST%
+	//   SELECT id, requestData, 'request' as type
+	//     FROM stored_requests
+	//     WHERE id in %REQUEST_ID_LIST%
+	//     UNION ALL
+	//   SELECT id, impData, 'imp' as type
+	//     FROM stored_imps
+	//     WHERE id in %IMP_ID_LIST%
 	//
 	// The MakeQuery function will transform this query into:
-	//   SELECT id, requestData FROM stored_requests WHERE id in ($1, $2, $3, ...)
+	//   SELECT id, requestData, 'request' as type
+	//     FROM stored_requests
+	//     WHERE id in ($1)
+	//     UNION ALL
+	//   SELECT id, impData, 'imp' as type
+	//     FROM stored_imps
+	//     WHERE id in ($2, $3, $4, ...)
 	//
 	// ... where the number of "$x" args depends on how many IDs are nested within the HTTP request.
 	QueryTemplate string `mapstructure:"query"`
-	// Hacky way to add Amp queries that won't fit in the global query above. A better solution would
-	// rename all the queries to be specific for what they are for. But that can wait for a general
-	// stored requests overhaul.
+
+	// AmpQueryTemplate is the same as QueryTemplate, but used in the `/openrtb2/amp` endpoint.
 	AmpQueryTemplate string `mapstructure:"amp_query"`
 }
 
-// MakeQuery gets a stored-request-fetching query which can be used to fetch numRequests requests at once.
-func (cfg *PostgresConfig) MakeQuery(numRequests int) (string, error) {
-	if numRequests < 1 {
-		return "", fmt.Errorf("can't generate query to fetch %d stored requests", numRequests)
-	}
-	return strings.Replace(cfg.QueryTemplate, "%ID_LIST%", makeIdList(numRequests), -1), nil
+// MakeQuery builds a query which can fetch numReqs Stored Requetss and numImps Stored Imps.
+// See the docs on PostgresConfig.QueryTemplate for a description of how it works.
+func (cfg *PostgresConfig) MakeQuery(numReqs int, numImps int) (query string) {
+	return resolve(cfg.QueryTemplate, numReqs, numImps)
 }
 
-// MakeAmpQuery gets a stored-request-fetching query which can be used to fetch numRequests requests at once.
-// A hacky solution for AMP, but a good solution would require an overhaul of the stored request system to
-// pass around the ID type. This should probably be its own PR.
-func (cfg *PostgresConfig) MakeAmpQuery(numRequests int) (string, error) {
-	if numRequests < 1 {
-		return "", fmt.Errorf("can't generate query to fetch %d stored requests", numRequests)
+// MakeAmpQuery is the equivalent of MakeQuery() for AMP.
+func (cfg *PostgresConfig) MakeAmpQuery(numReqs int, numImps int) string {
```

Comment:
Unless this is satisfying an interface, AMP requests take one request ID and zero IMP IDs, so the function needs not take any parameters.

Reply:
That's a good catch, but... yeah. It's not an interface, but there is a function-type argument. If I change it, this happens:

```
./pbs_light.go:961:51: cannot use cfg.Postgres.MakeAmpQuery (type func() string) as type func(int, int) string in argument to db_fetcher.NewFetcher
```